Move functions

zlobme opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/amer-on/cryptica-solver-java/3)
<!-- Reviewable:end -->

zlobme committed:
Move functions

Squashed branch

zlobme committed:
Merge 'move-through-target' into move1

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+
+        private int y;
+
+        private int x;
```
Думаю стоит поменять местами, все таки (x, y) более привычно чем (y, x).
Не забыть проверить все использования конструктора и поменять параметры местами.

Amer-On replied:
+

Amer-On replied:
+ done

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+
+    }
+
+    //    Move blocks
```
Раз хочешь писать комментарии, то:
* пиши везде
* пиши в формате javadoc

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+    //    Move blocks
+    public static char[][] move(char[][] gameField, String direction) {
+        List<Block> movableBlocks = findMovableBlocks(gameField);
+        List<Block> targetBlocks = findTargetBlocks(gameField);
```
unused, не понятно зачем это тут, лучше не пропускать такое в commit

Amer-On replied:
+

Amer-On replied:
+ плюс

Amer-On replied:
+ done

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+    public static char[][] move(char[][] gameField, String direction) {
+        List<Block> movableBlocks = findMovableBlocks(gameField);
+        List<Block> targetBlocks = findTargetBlocks(gameField);
+        if (direction.equals("left")) {
```
я бы заменил направления на enum

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+        } else if (direction.equals("down")) {
+            moveDown(gameField, movableBlocks);
+        }
+        return gameField;
```
Мы модифицируем входной параметр (что достаточно плохо), лучше делать копию, вопрос на подумать где? снаружи или внутри метода.
Опять же мы модифицируем `gamefield` внутри всех методов `moveXXX` и не принимаем возвращаемое значение, что тоже плохо.

Amer-On replied:
+ done

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+        for (Block block : movableBlocks) {
+            char newCell = gameField[block.y][block.x - 1];
+            if (newCell == '-') {
+                int newY = block.x - 1;
```
дублирование `block.x - 1` в трех местах, плохо... должно быть в одном месте.

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+                int newY = block.x - 1;
+                gameField[block.y][newY] = block.name;
+                gameField[block.y][block.x] = newCell;
+            } else if (Character.isUpperCase(newCell)) {
```
Если мы вынесли проверку и проставление `целей` в отдельный шаг алгоритма и метод, кажется что две ветки условий нам не нужны и мы всегда можем оставлять после себя пустое место.

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+
+    public static char[][] moveRight(char[][] gameField, List<Block> movableBlocks) {
+        movableBlocks.sort(Comparator.comparingInt(Block::getX));
+        Collections.reverse(movableBlocks);
```
Ну начать с того что 4 метода копипасты это оч плохо. Особенно что отличаются оно совсем чуть чуть в паре мест. Подумать как можно их унифицировать.

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+            System.out.print("Значение блока: " + block.name);
+            System.out.print(" Позиция по х: " + block.y);
+            System.out.println(" Позиция по у: " + block.x);
+        }
```
Выводить что-то в `sout` в реальной жизни вообще плохо, а еще и напрямую через `System.out` вообще харам. Вместо этого лучше завести методы форматирования, которые по входным параметрам будут возвращать отформатированную строку, готовую для вывода. И выводить ее отдельно куда мы захотим и через что мы захотим.

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+    }
+
+    // Finding blocks
+    public static List<Block> findMovableBlocks(char[][] gameField) {
+        List<Block> blocks = new ArrayList<Block>();
+        for (int i = 0; i < gameField.length; i++) {
+            for (int j = 0; j < gameField[i].length; j++) {
+                char cell = gameField[i][j];
+                if (cell != '-' && cell != '#' && !Character.isUpperCase(cell)) {
+                    Block block = new Block(cell, i, j);
+                    blocks.add(block);
+                }
+            }
+        }
+        return blocks;
+    }
+
+    public static List<Block> findTargetBlocks(char[][] gameField) {
+        List<Block> targetBlocks = new ArrayList<Block>();
+        for (int i = 0; i < gameField.length; i++) {
+            for (int j = 0; j < gameField[i].length; j++) {
+                char cell = gameField[i][j];
+                if (Character.isUpperCase(cell)) {
+                    Block targetBlock = new Block(cell, i, j);
+                    targetBlocks.add(targetBlock);
+                }
+            }
+        }
+        return targetBlocks;
+    }
```
2 метода близнеца, можно сократить до одного вынеся условие проверки в параметр `Predicate<Character> filter`

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+            for (int j = 0; j < gameField[i].length; j++) {
+                char cell = gameField[i][j];
+                if (cell != '-' && cell != '#' && !Character.isUpperCase(cell)) {
+                    Block block = new Block(cell, i, j);
```
К вопросу о перемене мест полей в классе `Block`, наверное стоит тут обозвать переменные x и y, чтобы еще больше сократить вероятность перепутать параметры конструктора местами.

Amer-On replied:
не совсем понял тебя, "х" и "у" вроде стоят как ты просил в одном из предыдущих коммитов
они стоят некрасиво со стороны математики, т.е. абсцисса с ординатой просто свитчнулись местами

Amer-On replied:
не совсем понял тебя, "х" и "у" вроде стоят как ты просил в одном из предыдущих коммитов
они стоят некрасиво со стороны математики, т.е. абсцисса с ординатой просто свитчнулись местами

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+    }
+
+    //    Print field out
+    public static void printGameField(char[][] gameField) {
+        for (char[] row : gameField) {
+            String separator = String.valueOf(" ");
+            String joined = CharBuffer.wrap(row).chars()
+                .mapToObj(intValue -> String.valueOf((char) intValue))
+                .collect(Collectors.joining(separator));
+            System.out.println(joined);
+        }
+    }
```
К вопросу о `sout`, лучше заменить все вызовы на вызов `formatGameField`, можно добавить параметр с разделителем для символов внутри строки.

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
     }
 
+    @Test
+    public void find_movable_blocks() {
```
Стоит добавить во все твои тесты проверки (`assertXXXX`), чтобы:
* понять что является правильным результатом
* отлавливать ошибки при написании и изменении кода

Сейчас они ничего не ловят.

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
+    }
+
+    @Test
+    public void move_left_right() {
```
Стоит разделять тестирование различных методов  по различным тестам.

Amer-On replied:
+ done

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
+        Main.move(gameField, "left");
+        Main.move(gameField, "left");
+
+        Main.checkTargetBlocksPos(gameField, targetBlocks);
```
Я оч против такого.. вызов метода должен быть внутри вызова `moveXXX`

Amer-On replied:
+ done

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
+        List<String> content = Arrays.asList("#--#", "#--#");
+        char[][] gameField = Main.fillGameField(content);
+        String result = Main.formatGameField(gameField);
+        System.out.println(result);
+        System.out.println(gameField);
```
У нас есть `assert`, лучше удалить эти отладочные выводы

Amer-On replied:
+ done

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
+    }
+
+    @Test
+    public void find_target_blocks() {
```
тот же случай, нет проверок

Amer-On replied:
+ done

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,107 @@ public void fill_game_field() {
+    }
+
+    @Test
+    public void check_target_block_pos() {
```
тот же случай, нет проверок

Amer-On replied:
+ done

zlobme reviewed (commented):
@Amer-On А еще я не могу переназначить "авторство" RP на тебя, так что лучше тебе их заводить.

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,164 @@
+        return gameField;
+    }
+
+    public static char[][] moveRight(char[][] gameField, List<Block> movableBlocks) {
```
```java
    @Test
    public void move_right_target() {
        List<String> content = Arrays.asList("#xxAx#");
        char[][] gameField = Main.fillGameField(content);
        var targetBlocks = Main.findTargetBlocks(gameField);

        Main.move(gameField, "right");

        Main.checkTargetBlocksPos(gameField, targetBlocks);
        String result = Main.formatGameField(gameField);
        assertThat(result).isEqualTo("#-xxx#");
    }
```
Я не стал добавлять тест, он падает с ошибкой
```
Expected :"#-xxx#"
Actual   :"#Axxx#"
```

Amer-On replied:
+ done

zlobme replied:
тест ты не добавил, а он так и падает, но уже с другой ошибкой
```
Expected :"#-xxx#"
Actual   :"#xxAx#"
```

Amer-On committed:
moving through target

Optimized moving through target. Added copyCharArray funct and changed all function's inside gameFields on copied arrays. Added tests and optimized some of them

Amer-On committed:
fixed mvn issues

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,183 @@
+    }
+
+    //    Move blocks
+    public static char[][] move(char[][] gameField,
```
вроде влезает в сроку, но если так переносишь, то переноси так
```
public static char[][] move(
    char[][] gameField,
    String direction,
    List<Block> targetBlocks
) {
```

Amer-On replied:
+ done

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,183 @@
+    }
+
+    public static char[][] moveLeft(char[][] gameField, List<Block> movableBlocks) {
+        movableBlocks.sort(Comparator.comparingInt(Block::getY));
```
ошибка, если мы хотим сдвинуться влево, но нужно смотреть самый левый блок, а не верхний

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,183 @@
+    }
 
+    public static char[][] moveRight(char[][] gameField, List<Block> movableBlocks) {
+        movableBlocks.sort(Comparator.comparingInt(Block::getY));
+        Collections.reverse(movableBlocks);
```
хотелось увидеть `movableBlocks.sort(comparingInt(Block::getX).reversed());`

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,183 @@
+        Collections.reverse(movableBlocks);
+        char[][] movedRightGameField = copyCharArray(gameField);
+        for (Block block : movableBlocks) {
+            char newCell = movedRightGameField[block.x][block.y + 1];
+            int newY = block.y + 1;
```
ожидалось, что будет 
```
int newY = block.y + 1;
char newCell = movedRightGameField[block.x][newY];
```

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -10,9 +12,183 @@
+    }
+
+    public static char[][] copyCharArray(char[][] charArray) {
+        char[][] charArrayCopy = new char[charArray.length][charArray[0].length];
+        for (int i = 0; i < charArray.length; i++) {
+            for (int j = 0; j < charArray[i].length; j++) {
+                charArrayCopy[i][j] = charArray[i][j];
+            }
+        }
+
+        return charArrayCopy;
```
Почти отлично, но так не делается (у тебя строки могут быть разной длинны, а мы это не проверяем), а то получишь `IndexOutOfBoundException`
```
char[][] charArrayCopy = new char[charArray.length][charArray[0].length];
```
есть лайфхак
```
public static char[][] copyCharArray(char[][] charArray) {
    char[][] result = new char[charArray.length][];
    System.arraycopy(charArray, 0, result, 0, charArray.length);
    return result;
}
```

Amer-On committed:
Update MainTest.java

zlobme commented on src/test/java/example/MainTest.java:
```diff
@@ -34,8 +30,125 @@ public void fill_game_field() {
+        List<String> content = Arrays.asList("#--A--#", "#--x--#");
+        char[][] gameField = Main.fillGameField(content);
+        char[][] newGameField = Main.copyCharArray(gameField);
+        newGameField[0][3] = 'a';
+        assertThat(newGameField).isNotEqualTo(gameField);
```
тут все не то :) надо проверить, что при копировании массивы
*  совпадают по содержимому
*  находятся по разным адресам

т.е.
```
assertThat(gameFieldCopy == gameField).isFalse();
assertThat(Arrays.equals(gameFieldCopy, gameField)).isTrue();
```
или в терминах assertj
```
assertThat(gameFieldCopy).isNotSameAs(gameField);
assertThat(gameFieldCopy).isEqualTo(gameField);
```

zlobme replied:
👍

Amer-On committed:
enum direction

made direction arg in move function enum type and optimized all tests

zlobme commented on src/main/java/example/Main.java:
```diff
@@ -26,8 +26,19 @@ public static void main(String[] args) {
 
     }
 
+    public static class enumValues {
+
+        enum direction {
+            left,
+            right,
+            up,
+            down
+        }
+
+    }
```
должно хватить
```
public static enum Directions {
    LEFT,
    RIGHT,
    UP,
    DOWN
}
```
и я бы вынес в отдельный файл и в отдельный файл класс `Block`

zlobme replied:
В - внимательность, странно что твой вариант прошел проверку

Amer-On committed:
direction -> enum

-changed direction type to enum and optimized everything for it
-optimized to be no errors in checkstyle
-made several unnecessary tests
